### ISSUE TITLE
Different behavior for registering an interface as an object vs a function

### ISSUE THREAD

**Issue by madhuravi** [uber-go/dig#27]
Dig behaves differently when registering an interface as an object vs a function.

Eg.
type Scope interface {}
type scopeStruct struct {}
var s Scope = scopeStruct{}
1. As object:
dig.MustRegister(s)
dig.MustResolve(&s) -> dependency of type Scope is not registered
2. As function:
dig.MustRegister(func() (Scope, error) {return scope,nil})
dig.MustResolve(&s) -> works fine

In case 1, dig registers it as a *scopeStruct struct instead of Scope interface.

Is this the behavior we want from dig? It would be great if both are consistent and dig.Resolve works in case 1 as well but seems harder to do.

**Comment by glibsm** [comment]
@madhuravi are you sure you used the latest dig? It's something that was fixed post v0.1 by @dmcaulay.

```
package main

import (
	"fmt"

	"go.uber.org/dig"
)

type Scope interface{}
type scopeStruct struct{}

var s Scope = scopeStruct{}

func main() {
	c := dig.New()

	c.MustRegister(&s)
	fmt.Println(c)

	var r Scope
	c.MustResolve(&r)
	fmt.Printf("%#v", r)
}

```

correctly shows 

```
{nodes:
main.Scope -> (object) obj: main.Scope, deps: nil, cached: true, cachedValue: {}
}

main.scopeStruct{}
```

**Comment by glibsm** [comment]
Particular commit in question: https://github.com/uber-go/dig/commit/117d65327373a1805b00ca98a74ae8537f8a83d2.

Looks like at least v0.2 of dig is required to get the functionality.
